### DISCUSSION TITLE
compatibility with python=3.8 or python=3.9

### DISCUSSION THREAD

**Discussion by r-shruthi11** [talmolab/sleap#1102] [Help!] [answered]
Hey yall,

`sleap=1.2.9` is supposed to be compatible with `python=3.8` but conda complains when I try to install it from the sleap channel into an environment that contains `python=3.8`

```
conda install -c sleap sleap
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: \
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - sleap -> python[version='3.6.*|3.7.*']

Your python: python=3.8.0

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that.
```

**Comment by roomrys** [comment, marked as answer]
Hi @r-shruthi11,

The conda package requires `python=3.7`. If you need python 3.8, then you can try to pip install sleap from within the conda environment.

Let us know if you run into any trouble.

Thanks,
Liezl

**Reply by r-shruthi11** [reply]
Hi Liezl,

Thanks for getting back to me. When I pip install `sleap` into an environment with `python=3.8` and `cudatoolkit` and `cudnn` it doesn't install with GPU support. Has this been your experience as well? How do I ensure that I install the appropriate sleap version with GPU support? 

Shruthi

**Reply by r-shruthi11** [reply]
Solved it! Thanks for your help.

**Reply by r-shruthi11** [reply]
For posterity: It helps to install cudatoolkit and cudnn with python while setting up the conda environment 
`conda create -n my_env -c conda-forge cudatoolkit cudnn python=3.8` and then `pip install sleap`.

**Reply by talmo** [reply]
Yeah, those are kind of tricky -- the cudatoolkit and cudnn versions need to align with TensorFlow's.

For reference, we use [TensorFlow 2.6, cudatoolkit 11.3, cudnn 8.2](https://sleap.ai/help.html#what-if-i-already-have-cuda-set-up-on-my-system) by default.

For completeness:
```
conda create -n sleap python=3.8 cudatoolkit=11.3 cudnn=8.2 && conda activate sleap && pip install sleap
```

You can also `git clone https://github.com/talmolab/sleap && cd sleap`, then edit `environment.yml` to use `python=3.8`, and install it with: `conda env create -f environment.yml`

In the future we will be upgrading the default conda package to use 3.8 and may consider packaging multiple versions if needed -- there's just a lot of dependencies that need to get updated for full cross-OS support 😢

**Comment by niksirbi** [comment]
Hey SLEAP developers, thanks for the awesome work you are doing!

I'd like to ask whether you are planning to support newer Python versions in upcoming releases.
All the conda environment files still mandate `python=3.7`, except for the apple-silicon one (`python 3.9`).

The reason I'm asking is the following:
I am in the early phases of developing a [Python package for analysing animal behaviour](https://github.com/neuroinformatics-unit/movement). I really like the SLEAP data model and I want to try adopting it (os subclassing it) in my package. However, to do that I'd have to include SLEAP as a dependency. Since my own package requires `python > 3.8`, I have troubles installing SLEAP alongside it, as the dependencies cannot resolve.

I presume tensorflow + cuda make it very hard to support `Python > 3.8` in a way that the whole dependency stack works. That said, I don't really need tensorflow/cuda for my usecase, since I'm mainly interested in using the SLEAP classes such as `Animal`, `Point`, `Trajectory`, etc.

Would it be possible to provide a pip installation option with a subset of core dependencies, say something like `pip install sleap[minimal]`? 

I guess the alternative for me would be to vendor some of the relevant SLEAP modules into my package?

Interested in hearing your thoughts on this. Maybe there is already a workaround I haven't thought of.

Niko

**Reply by talmo** [reply]
Hi @niksirbi,

Cool package! We'd love to help you support SLEAP data.

Indeed, TensorFlow + cuda are the exact blockers for us right now. Technically they'll work if we update Python, but due to #868 → https://github.com/tensorflow/tensorflow/issues/52988, we're a bit blocked. Additionally, in newer versions of TensorFlow that have binaries for other Python versions, Google is dropping native Windows support in favor of WSL, which we really don't think will serve our users well.

All that said, you're totally right that you don't need all of this for your package. This is a common scenario for downstream developers that mostly just need to be able to interact with the data formats/structures. To that end, we've been working on [sleap-io](https://github.com/talmolab/sleap-io). You can install it via `pip install sleap-io` on any platform and Python version (with 3.9 being the default). It has minimal dependencies and we've reimplemented the data model and deserialization routines. It's not as feature-rich as SLEAP, but we've been using it without major issues in some internal projects successfully so far. We don't have docs generated for that yet, but it's a pretty minimal package and the docstrings are pretty exhaustive.

Let us know if that package fits the bill for you and feel free to open issues on the sleap-io repo if there's anything missing that you need or that isn't working!

Cheers,

Talmo

**Reply by niksirbi** [reply]
Thanks for the speedy and super helpful response @talmo!

I understand TensorFlow is a pain for this and I hope you find a way out of this bind eventually.

`sleep.io` looks exactly like the thing we need! Thanks for pointing me to it, this will save us lots of work. We'll try to use that and open issues on its repository if we encounter problems.

Cheers,
Niko
